src/thermo/PlasmaPhase.cpp

```diff
@@ -0,0 +1,107 @@
+    setElectronTemperature(temperature());
+}
+
+void PlasmaPhase::updateIsotropicElectronEnergyDistribution()
+{
+    m_electronEnergyDistrb.resize(m_nPoints);
+    double gamma1 = boost::math::tgamma(3.0 / 2.0 * m_x);
+    double gamma2 = boost::math::tgamma(5.0 / 2.0 * m_x);
+    double c1 = m_x * std::pow(gamma2, 1.5) / std::pow(gamma1, 2.5);
+    double c2 = m_x * std::pow(gamma2 / gamma1, m_x);
+    m_electronEnergyDistrb =
+        c1 * m_electronEnergyLevels.array().sqrt() /
+        std::pow(m_meanElectronEnergy, 1.5) *
+        (-c2 * (m_electronEnergyLevels.array() /
+        m_meanElectronEnergy).pow(m_x)).exp();
+}
+
+void PlasmaPhase::setElectronTemperature(const double Te) {
+    Phase::setElectronTemperature(Te);
+    m_meanElectronEnergy = 3.0 / 2.0 * electronTemperature() *
+                           GasConstant / (Avogadro * ElectronCharge);
+    updateIsotropicElectronEnergyDistribution();
+}
+
+void PlasmaPhase::setElectronEnergyLevels(const vector_fp& levels)
+{
+    m_nPoints = levels.size();
+    m_electronEnergyLevels =
+        Eigen::Map<const Eigen::VectorXd> (levels.data(), levels.size());
+    updateIsotropicElectronEnergyDistribution();
+}
+
+void PlasmaPhase::getElectronEnergyLevels(vector_fp& levels) const
+{
+    levels.resize(m_nPoints);
+    Eigen::Map<Eigen::VectorXd> (levels.data(), m_nPoints) = m_electronEnergyLevels;
+}
+
+void PlasmaPhase::setElectronEnergyDistribution(const vector_fp& levels,
+                                                const vector_fp& distrb)
+{
+    if (levels.size() != distrb.size()) {
+        throw CanteraError("PlasmaPhase::setElectronEnergyDistribution",
+                           "Vector lengths need to be the same.");
+    }
+    setElectronEnergyLevels(levels);
+    m_nPoints = levels.size();
+    m_electronEnergyDistrb =
+        Eigen::Map<const Eigen::VectorXd> (distrb.data(), distrb.size());
+    // calculate mean electron energy and electron temperature
+    Eigen::VectorXd eps52 = m_electronEnergyLevels.array().pow(5./2.);
+    m_meanElectronEnergy =
+        2.0 / 5.0 * trapezoidal(m_electronEnergyDistrb, eps52);
+    Phase::setElectronTemperature(
+        2.0 / 3.0 * m_meanElectronEnergy * Avogadro *
+        ElectronCharge / GasConstant);
+}
+
+void PlasmaPhase::getElectronEnergyDistribution(vector_fp& distrb) const
+{
+    distrb.resize(m_nPoints);
+    Eigen::Map<Eigen::VectorXd> (distrb.data(), m_nPoints) = m_electronEnergyDistrb;
+}
+
+void PlasmaPhase::updateThermo() const
+{
+    IdealGasPhase::updateThermo();
+    static const int cacheId = m_cache.getId();
+    CachedScalar cached = m_cache.getScalar(cacheId);
+    double tempNow = temperature();
+    double electronTempNow = electronTemperature();
+    size_t k = speciesIndex(m_electronName);
+    // If the electron temperature has changed since the last time these
+    // properties were computed, recompute them.
+    if (cached.state1 != tempNow && cached.state2 != electronTempNow) {
```

Comment:
I think checking gas temperature here is redundant.

Reply:
I think you need to be more careful with the update here. Any time the gas temperature changes, it will replace the value of `m_cp0_R` for the electron species with a value calculated as if its temperature was the gas temperature. So in that case, you also need to re-do the calculation using the actual electron temperature to overwrite that value.

Reply:
I think the condition here should be "or", not "and" -- you need to recalculate the electron thermo if either temperature has changed.